x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -511,4 +516,30 @@ private void updateEnrichPolicyAlias(final String destinationIndexName) {
     private Client enrichOriginClient() {
         return new OriginSettingClient(client, ENRICH_ORIGIN);
     }
+
+    private static Client wrapClient(Client in, String policyName, ExecuteEnrichPolicyTask task, ClusterService clusterService) {
+        // Filter client in order to:
+        // 1) Check on transport action call that policy runner does whether the task has been cancelled
+        // 2) Set the enrich policy task as parent task, so if other API calls (e.g. reindex) are cancellable then
+        // the corresponding tasks of these API calls get cancelled as well.
+        return new FilterClient(in) {
+
+            @Override
+            protected <Request extends ActionRequest, Response extends ActionResponse> void doExecute(
+                ActionType<Response> action,
+                Request request,
+                ActionListener<Response> listener
+            ) {
+                String requestStep = request.getClass().getSimpleName();
+                task.setStep(requestStep);
+                if (task.isCancelled()) {
+                    String message = "cancelled policy execution [" + policyName + "], status [" + Strings.toString(task.getStatus()) + "]";
+                    listener.onFailure(new TaskCancelledException(message));
+                    return;
+                }
```

Comment:
There isn't a way to propagate a task cancellation to its child tasks by default is there? Just asking for my own knowledge here, not suggesting any changes.

Reply:
Via this if clause, the node task let's any thread that continues execution know that the node task has been cancelled and the on failure invocation triggers info log in InternalExecutePolicyAction line 121. The `TaskCancellationService#doCancelTaskAndDescendants(...)` method handles cancellation of child tasks `(which in case of `ExecuteEnrichPolicyTask` any cancellable task backed by transport action calls invoked from the wrapped client in `EnrichPolicyRunner`(e.g. reindex) ). The `TaskCancellationService` knows about this because in line 540 of `EnrichPolicyRunner` class the ExecuteEnrichPolicyTask's id is set as parent task.

Reply:
Thanks for that explanation!